### ISSUE TITLE
Msg objects that reuse "tcp request" sometimes have unexpected output

### ISSUE THREAD

**Issue by fullmooooon** [node-red/node-red#4378]
### Current Behavior

If i use `msg = { payload: [] }`  
Everything is fine.
But if I use `msg.payload = []`
"split" is starting to output tcp buffer data!
It's like getting the next set of tcp buffer data.

```
; (() => {
    let i = 0
    while (i < 100000000) i++
})()
```
In my real code, there is some buffer processing that is not so time-consuming. 
So it happens once in a while, every ten seconds. 
I replaced the sample code with the more time-consuming loop code.

<img width="762" alt="image" src="https://github.com/node-red/node-red/assets/43309591/89b523d0-263b-43aa-b1a4-1ab47c8c61cb">

### Steps To Reproduce

Tcp connects to a server that sends less than ten buffer data of no more than 2000 in length per second.

### Example flow

```
[{"id":"fb257b355f23da96","type":"inject","z":"8616f41adb0d8dd9","name":"","props":[{"p":"payload"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","payload":"","payloadType":"date","x":510,"y":720,"wires":[["1cc529270b3cd233"]]},{"id":"0a92104b96294e55","type":"function","z":"8616f41adb0d8dd9","name":"Message processing","func":"// msg = { payload: [] }\nmsg.payload = []\n; (() => {\n    let i = 0\n    while (i < 100000000) i++\n})()\nreturn msg;","outputs":1,"timeout":0,"noerr":0,"initialize":"","finalize":"","libs":[{"var":"fs","module":"fs"}],"x":940,"y":720,"wires":[["a2614a53111442cd"]]},{"id":"1cc529270b3cd233","type":"tcp request","z":"8616f41adb0d8dd9","name":"","server":"10.168.1.115","port":"6789","out":"sit","ret":"buffer","splitc":" ","newline":"","trim":false,"tls":"","x":700,"y":720,"wires":[["0a92104b96294e55"]]},{"id":"a2614a53111442cd","type":"debug","z":"8616f41adb0d8dd9","name":"debug 26","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"payload","targetType":"msg","statusVal":"","statusType":"auto","x":1140,"y":720,"wires":[]}]
```

### Environment

- Node-RED version: 3.1.0
- Node.js version: v20.6.1
- npm version: 8.19.1
- Platform/OS: win11
- Browser: Edge 117.0.2045.55

**Comment by Steve-Mcl** [comment]
Hi, i have tried your demo flow in NR 3.10 and cannot see what you see.

You mention the "split" node in your OP ""split" is starting to output tcp buffer data!" but there is no split node in your flows.

Please try to create a demo flow that reliably recreates your issue.

**Comment by fullmooooon** [comment]
> Hi, i have tried your demo flow in NR 3.10 and cannot see what you see.
> 
> You mention the "split" node in your OP ""split" is starting to output tcp buffer data!" but there is no split node in your flows.
> 
> Please try to create a demo flow that reliably recreates your issue.

I may have encountered some problems with node-red "caching". 
I suspect that it causes my node-red code to show an inconsistency between the flow shown and the actual run. 
The key to trigger is whether to reuse msg objects. 
Some tcp data streams seem to skip the "Message processing" Function in my example.
I did not change any code, after two days now, and then the inexplicable output disappeared, and now my code is fine. 
So I decided to turn off this issue. 
But the confusion has not gone away. 
If I have any other findings about errors, I will report.

**Comment by fullmooooon** [comment]
@Steve-Mcl
Hi, my node-red has this problem again two hours ago.
but until now I had a meal and it was back to normal.🤨
I suspect it has something to do with an exe software written in python that I have just used, and my job is to try to implement part of the software's functions in node-red. 
So the software will also connect to the tcp server after it is opened. 
I don't know what stupid thing this software did after it was shut close, Maybe the tcp connection was put on hold. 
Then the tcp-related code in node-red enters this abnormal state.
I would like to describe what happened in this "wrong state". 
Just like the code in Example flow, if you reuse msg objects, some tcp data will skip "Message processing" to "debug 26". 
Maybe it was replaced with the next tcp buffer data before reaching "debug 26"? 
In addition, in my actual code, the data received by the Function after "tcp request" will be out of order or lost, even if the msg is a new "{payload:object}". 
I don't know which one is missing or the wrong sequence.

I have tried to restart and upgrade node and win11, and node-red will not quit this state, except waiting.
I think I must repeat that I am not joking. 
Does anyone have any idea?

**Comment by fullmooooon** [comment]
@Steve-Mcl I've tried it many times.
If I change line 649 of "node-red\ node_modules\ @ node-red\ nodes\ core\ network\ 31-tcpin.js"
<img width="466" alt="image" src="https://github.com/node-red/node-red/assets/43309591/7e29d60e-3339-4a26-8bb7-2a3dec79097b">
Creating a new msg object all the time will not output buffer.
It seems that tcp msg is reassigned to the next data before the debug output.
I confirmed in "Wireshark" that the order of TCP is fine.

OK, in short, I will organize the proof to go to the forum for help.

**Comment by Steve-Mcl** [comment]
@fullmooooon 

If you believe this is definitely a problem in Node-RED and you can provide a working proof, then here is the right place to continue the discussion.

The forum is often used to discover the problem (to triage it) but ultimately, an issue is raised once it it confirmed.

PS: there are are some very helpful folk on the forum and it is far more active there than here on the Github issues pages.

**Comment by gavrilyak** [comment]
I fought with this issue too, and it is definitely a problem in Node-RED. 
It is not so simple to provide a working proof, but the  change proposed in [message above](https://github.com/node-red/node-red/issues/4378#issuecomment-1754678860) definitely fixes  issue for me. 

It is definitely a bug in Node-RED, as the code tries to use `lastMsg`, but `lastMsg` is assigned only  in line 602 and this assignment is an INPUT message to node, e.g. the data we want to **send** to remote, this `lastMsg` has nothing to do with the data we are **receiving** in on('data') handler on line 683.  That's why is a just bug, no proof needed, just code review. 

And there are more usages of `lastMsg` in on('data') handler,  they are suspicious too...

Sorry for commenting on closed issue and for my English and tone, I am not a native speaker.
Please reopen this issue.

**Comment by fullmooooon** [comment]
@gavrilyak 

Node-RED always efficiently reuses objects instead of creating new ones. During execution, it also doesn’t run in the order you might assume. You must immediately store the current data again in the next node after the TCP node.

🤔The maintainer doesn't consider this to be a problem, but it has now become a must-face challenge for everyone using TCP nodes.

> we need to create a new msg object in the first function that follows "tcp request"

https://discourse.nodered.org/t/tcp-msg-is-reassigned-to-the-next-data-before-debug-output/81907/18#:~:text=fundamental%20problem%2C%20and-,we%20need%20to%20create%20a%20new%20msg%20object%20in%20the%20first%20function%20that%20follows%20%22tcp%20request%22,-.At%20this%20point

**Comment by gavrilyak** [comment]
Yes, Node-RED is good with reusing object. That problem is not with Node-RED,  only with tcp node, but it a "core" node. This node reuses wrong message object and sometimes passes input data directly to the output.
 
The problem was introduced in [this commit](https://github.com/node-red/node-red/commit/db1b0ccb79ee74d3e72ec3ea8fe16e11df77ca50). 

Repeatably reproducing conditions that leads to the problem to give a proof is hard, very specific conditions needed that depends on OS, network speed, network buffers etc. It  looks like random "magic" that sometimes just happens.
That's why I suggested to  look at the actual code as the easiest way to see what this issue is about. 

To reiterate: `lastMsg` must not be reused in receiving part, that commit was about transmit, but somehow  touched the receive part.

**Comment by fullmooooon** [comment]
@gavrilyak I’m no longer focusing on this issue. You can open a new one if needed.

**Comment by gavrilyak** [comment]
I'm too not interested in this node, as it is not suitable for my use case anyway. I just wanted to provide more info for future users who will stumble with strange behavior of this node and probably find  this ticket as I found it. Thank you for your insights, they helped me to understand what  'magic'  was really  really going.
